### PR TITLE
Enhanced Secret Functionality

### PR DESCRIPTION
**Pull request by ejether**
This PR adds a new key to the course and new functionality for secret handling. The `secret` top level key is a list of maps that define how and hwere to get a secret value. The secrets are merged into the `os.environ` and that is used as before. Duplicate variables are not allowed. 

This pr adds two providers `AWSParameterStore` and `ShellExecutor`

### CODE REVIEW

**Comment by ejether** [comment]
I still need to add documentation and tests for the `ShellExecutor` but I wanted to get this out there sooner. 
I'd love feedback on the code structure, if there is any, before I start to go to town on the documentation. 
The way the secret provider frame is written feels like it might be wonky. with `SecretProvider.get_value()` not accepting arguments. 🤷 

All in all, this was really fun to write and it came together really nicely.

**Comment by ejether** [comment]
Closes: https://github.com/FairwindsOps/reckoner/issues/268

**Comment by ejether** [comment]
@pthomison you might be interested in this as well.

File: end_to_end_testing/course_files/13_test_lint_bad_secret.yml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+namespace: "doesnotmatter"
+repository: "stable"
+repositories:
+  stable: "this is invalid schema"
+charts:
+  my-chart:
+    chart: nginx-ingress
+secrets:
+- backend: testbackend
+- name: testname
```
Comment (ejether) [inline end_to_end_testing/course_files/13_test_lint_bad_secret.yml:10]:
```suggestion
- name: testname

```

File: end_to_end_testing/course_files/13_test_lint_good_secret.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+repository: "stable"
+repositories:
+  stable: "this is invalid schema"
+charts:
+  my-chart:
+    chart: nginx-ingress
+secrets:
+- backend: testbackend
+  name: testname
+  allowed: test_other_allowed
```
Comment (ejether) [inline end_to_end_testing/course_files/13_test_lint_good_secret.yml:11]:
```suggestion
  allowed: test_other_allowed
  
```

File: end_to_end_testing/tests/13_schema.yaml [hunk 3]
```diff
@@ -17,3 +17,20 @@ testcases:
...
+    - result.systemout ShouldContainString '\'backend\' is a required property'
+    - result.systemout ShouldContainString '\'name\' is a required property.'
+
+- name: 13 - plot 13_test_lint_good_secret.yml
+  steps:
+  - script: |
+      reckoner lint ../course_files/13_test_lint_good_secret.yml
+    assertios:
+    - result.code ShouldEqual 0
+    - result.systemout ShouldContainString 'No schema validation errors found'
```
Comment (ejether) [inline end_to_end_testing/tests/13_schema.yaml:36]:
```suggestion
    - result.systemout ShouldContainString 'No schema validation errors found'

```

File: reckoner/tests/test_course.py [hunk 4]
```diff
@@ -153,6 +154,58 @@ def test_hooks_parsed(self, mockHook, mockGetHelm, mockYAML):
         self.assertEqual(course.pre_install_hook.commands, ['command1', 'command2'])
         self.assertEqual(course.post_install_hook.commands, ['command3', 'command4'])
 
+    def test_secrets_parsed(self, mockHook, mockGetHelm, mockYAML):
+
+        course_with_secrets = self.course_yaml.copy()
+        # Doing this because the way secretes get injected into the environment
```
Comment (sudermanjr) [inline reckoner/tests/test_course.py:160]:
```suggestion
        # Doing this because the way secrets get injected into the environment
```

**Review by sudermanjr — CHANGES_REQUESTED** [review]
Seems good. Haven't tested myself, but the code looks good. I didn't notice, is there an e2e test that uses the shell executor to test that code path?

File: docs/README.md [hunk 5]
```diff
@@ -179,6 +191,28 @@ namespace_management:
       overwrite: True
 ```
 
+## Secrets
+  A list of maps that define how and where to get secret values to inject into the environment.
+
```
Comment (ejether) [inline docs/README.md:196]:
```suggestion
```

File: reckoner/course.py [hunk 6]
```diff
@@ -169,8 +181,31 @@ def pre_install_hook(self):
...
+        Accepts no Argument
+        Returns None
+
+        Loops over list of secrets and merges the name:values into the environment
+        Throws ReckonerException if there is an existing Environment of the same name
+        """
+
+        for secret in self.secrets:
+            if secret.name in os.environ.keys():
+                raise ReckonerException(
```
[the remark this answers is not in the document]
Comment (ejether) [reply reckoner/course.py:195]: Agreed, we will need some optimization if we add secrets at the chart and the global scope. 
This was the simplest and least disruptive way I could think of to accomplish the global scope.

**Comment by ejether** [comment]
There is not an e2e test for that. I'll add one.

**Review by sudermanjr — APPROVED** [review]
I am happy with this as it is. The structure seems sufficient for MVP, and it will solve most use-cases. Waiting for review from Luke before I merge.

**Review by lucasreed — APPROVED** [review]

**Pull request merged**